File: src/examples/cpp/protobuf_rpc/module/normal_rpc_server_module/normal_rpc_server_module.cc
#include "normal_rpc_server_module/normal_rpc_server_module.h"
#include "normal_rpc_server_module/filter.h"
#include "normal_rpc_server_module/global.h"

namespace aimrt::examples::cpp::protobuf_rpc::normal_rpc_server_module {

bool NormalRpcServerModule::Initialize(aimrt::CoreRef core) noexcept {
  core_ = core;

  SetLogger(core_.GetLogger());

  try {
    // 注册rpc服务
    service_ptr_ = std::make_shared<ExampleServiceImpl>();

    // Register filter
    service_ptr_->RegisterFilter(DebugLogServerFilter);
    service_ptr_->RegisterFilter(TimeCostLogServerFilter);

    bool ret = core_.GetRpcHandle().RegisterService(service_ptr_.get());
    AIMRT_CHECK_ERROR_THROW(ret, "Register service failed.");

    AIMRT_INFO("Register service succeeded.");

  } catch (const std::exception& e) {
    AIMRT_ERROR("Init failed, {}", e.what());
    return false;
  }

  AIMRT_INFO("Init succeeded.");

  return true;
}

bool NormalRpcServerModule::Start() noexcept { return true; }

void NormalRpcServerModule::Shutdown() noexcept {}

}  // namespace aimrt::examples::cpp::protobuf_rpc::normal_rpc_server_module

File: src/examples/py/protobuf_rpc/py/normal_rpc_server_module.py
import aimrt_py
import aimrt_py_log

import rpc_pb2
import rpc_aimrt_rpc_pb2

from google.protobuf.json_format import MessageToJson


class ExampleServiceImpl(rpc_aimrt_rpc_pb2.ExampleService):
    def __init__(self, logger):
        super().__init__()
        self.logger = logger

    def GetFooData(self, ctx_ref, req):
        rsp = rpc_pb2.GetFooDataRsp()
        rsp.msg = "echo " + req.msg

        aimrt_py_log.info(self.logger,
                          "Server handle new rpc call. req: {}, return rsp: {}"
                          .format(MessageToJson(req), MessageToJson(rsp)))

        return aimrt_py.RpcStatus(), rsp

    def GetBarData(self, ctx_ref, req):
        rsp = rpc_pb2.GetBarDataRsp()
        rsp.msg = "echo " + req.msg

        aimrt_py_log.info(self.logger,
                          "Server handle new rpc call. req: {}, return rsp: {}"
                          .format(MessageToJson(req), MessageToJson(rsp)))

        return aimrt_py.RpcStatus(), rsp


class NormalRpcServerModule(aimrt_py.ModuleBase):
    def __init__(self):
        super().__init__()
        self.core = aimrt_py.CoreRef()
        self.logger = aimrt_py.LoggerRef()

    def Info(self):
        info = aimrt_py.ModuleInfo()
        info.name = "NormalRpcServerModule"
        return info

    def Initialize(self, core):
        self.core = core
        self.logger = self.core.GetLogger()

        # log
        aimrt_py_log.info(self.logger, "Module initialize")

        try:
            # rpc-server
            self.service = ExampleServiceImpl(self.logger)
            ret = self.core.GetRpcHandle().RegisterService(self.service)
            if (not ret):
                aimrt_py_log.error(self.logger, "Register service failed.")
                return False

        except Exception as e:
            aimrt_py_log.error(self.logger, "Initialize failed. {}".format(e))
            return False

        return True

    def Start(self):
        aimrt_py_log.info(self.logger, "Module start")

        return True

    def Shutdown(self):
        aimrt_py_log.info(self.logger, "Module shutdown")
